File: LC 93.cpp
class Solution {
public:
    vector<string> restoreIpAddresses(string s) {
        vector<string> ans;
        int n = s.length();
        string ip = "";
        function<void(int, int)> dfs = [&] (int idx, int num) {
            if (idx >= n) {
                if (num == 4) {
                    ip.pop_back();
                    ans.push_back(ip);
                }
                return;
            }
            int t = 0;
            for (int i = idx; i < n; i++) {
                t = t * 10 + int(s[i] - '0');
                if (num > 3 || t > 255 || (i > idx && s[idx] =='0')) break;
                string ss = ip;
                ip += to_string(t) + ".";
                dfs(i + 1, num + 1);
                ip = ss;
            }
        };
        dfs(0, 0);
        return ans;
    }
};

File: 1000/LC 93.go
func restoreIpAddresses(s string) (ans []string) {
    path := []byte{}
    n := len(s)

    var dfs func(int, int)
    dfs = func(idx, cnt int) {
        if idx == n && cnt == 4 {
            ans = append(ans, string(path[:len(path)-1]))
        } else if idx >= n || cnt >= 4 {
            return
        } 

        t, m := 0, len(path) 
        for i := idx; i < n; i++ {
            t  = t * 10 + int(s[i] - '0')
            if i > idx && s[idx] == '0' || t > 255{
                break
            }
            path = append(path, s[i])
            path = append(path, '.')
            dfs(i + 1, cnt + 1)
            path = path[:len(path)-1]
        }
        path = path[:m]
    }
    dfs(0, 0)

    return 
}
